migrations/migration.go

```diff
@@ -559,6 +449,262 @@ func (m *StorageMigration) MigrateNestedValue(
+		) {
+			newValue := m.MigrateNestedValue(
+				storageKey,
+				storageMapKey,
+				existingValue,
+				valueMigrations,
+				reporter,
+				allowMutation,
+			)
+
+			var valueToSet interpreter.Value
+			if newValue == nil {
+				valueToSet = existingValue
+			} else {
+				valueToSet = newValue
+
+				// Remove existing value since value is migrated.
+				existingValue.DeepRemove(inter)
+				inter.RemoveReferencedSlab(existingValueStorable)
+			}
+
+			owner := dictionary.GetOwner()
+
+			pathDomain := common.PathDomainStorage
+
+			storageMap := m.storage.GetStorageMap(owner, pathDomain.Identifier(), true)
+			conflictDictionary := interpreter.NewDictionaryValueWithAddress(
+				inter,
+				emptyLocationRange,
+				dictionary.Type,
+				owner,
+			)
+			conflictDictionary.InsertWithoutTransfer(
+				inter,
+				emptyLocationRange,
+				newKey,
+				valueToSet,
+			)
+
+			conflictStorageMapKey := m.nextDictionaryKeyConflictStorageMapKey()
+
+			addressPath := interpreter.AddressPath{
+				Address: owner,
+				Path: interpreter.PathValue{
+					Domain:     pathDomain,
+					Identifier: string(conflictStorageMapKey),
+				},
+			}
+
+			if storageMap.ValueExists(conflictStorageMapKey) {
+				panic(errors.NewUnexpectedError(
+					"conflict storage map key already exists: %s", addressPath,
+				))
+			}
+
+			storageMap.SetValue(
+				inter,
+				conflictStorageMapKey,
+				conflictDictionary,
+			)
+
+			reporter.DictionaryKeyConflict(addressPath)
+
+		} else {
+
+			// No conflict, insert the new key and existing value pair
+			// Don't migrate value here because we are going to migrate all values in the dictionary next.
+
+			dictionary.InsertWithoutTransfer(
+				inter,
+				emptyLocationRange,
+				newKey,
+				existingValue,
+			)
+		}
+	}
+}
+
+func (m *StorageMigration) migrateDictionaryValues(
+	storageKey interpreter.StorageKey,
+	storageMapKey interpreter.StorageMapKey,
+	dictionary *interpreter.DictionaryValue,
+	valueMigrations []ValueMigration,
+	reporter Reporter,
+	allowMutation bool,
+) {
+
+	inter := m.interpreter
+
+	type keyValuePair struct {
+		key, value interpreter.Value
+	}
+
+	var existingKeysAndValues []keyValuePair
+
+	dictionary.Iterate(
+		inter,
+		func(key, value interpreter.Value) (resume bool) {
+
+			existingKeysAndValues = append(
+				existingKeysAndValues,
+				keyValuePair{
+					key:   key,
+					value: value,
+				},
+			)
+
+			// Continue iteration
+			return true
+		},
+		emptyLocationRange,
+	)
+
+	for _, existingKeyAndValue := range existingKeysAndValues {
+		existingKey := existingKeyAndValue.key
+		existingValue := existingKeyAndValue.value
+
+		newValue := m.MigrateNestedValue(
+			storageKey,
+			storageMapKey,
+			existingValue,
+			valueMigrations,
+			reporter,
+			allowMutation,
+		)
+
+		if newValue == nil {
+			continue
+		}
+
+		if !allowMutation {
+			panic(errors.NewUnexpectedError(
+				"mutation not allowed: attempting to migrate dictionary value: %s",
+				existingValue,
+			))
+		}
```

Comment:
same here: could take it out of the loop, since it will report only one error anyway (exits at the first error)

Reply:
We should only check if we're allowed to mutate if we actually are going to mutate, i.e. if `newValue != nil`. It might be the case that none of the values need to be migrated, in which case we should not panic with an error that we're not allowed to mutate

Reply:
ah that makes sense!

Reply:
Great question though! I added comments in 13e8e6b1193b02a172b64e2e021f96568aeaf44a